lib/Sema/CodeSynthesis.cpp

```diff
@@ -1643,6 +1646,55 @@ void synthesizeAccessorBody(AbstractFunctionDecl *fn, void *) {
   llvm_unreachable("bad synthesized function kind");
 }
 
+static void maybeAddMemberwiseDefaultArg(ParamDecl *arg, VarDecl *var,
+                    SmallVectorImpl<DefaultArgumentInitializer *> &defaultInits,
+                                         unsigned paramSize, ASTContext &ctx) {
+  // First and foremost, if this is a constant don't bother.
+  if (var->isLet())
+    return;
+
+  // We can only provide default values for patterns binding a single variable.
+  // i.e. var (a, b) = getSomeTuple() is not allowed.
+  if (!var->getParentPattern()->getSingleVar())
+    return;
+
+  // In batch mode, lazy variables may or not have had their accessors
+  // synthesized yet. In the case that they have, we still want to emit a
+  // default arg.
+  auto areLazyAccessorsSynthesized = !var->hasStorage() &&
+                                     var->getAttrs().hasAttribute<LazyAttr>();
+
+  // If we don't have an expression initializer or silgen can't assign a default
+  // initializer, then we can't generate a default value. An example of where
+  // silgen can assign a default is var x: Int? where the default is nil.
+  if (!areLazyAccessorsSynthesized &&
```

Comment:
I'm worried that this adds some non-determinism to the behavior here. There's no guarantee that the lazy property has been completed at the time you're synthesizing the default argument for `var`.

However, you really don't care one way or another: whether or not the lazy property's storage has been synthesized should not affect the behavior here. In fact, the type of a lazy property parameter to a memberwise init is an Optional of the lazy property's type, therefore the default argument value for a lazy property is the default argument value for the lazy property's _storage_, which is always nil. So you can remove the areLazyAccessorsSynthesized and instead treat the default argument as a nil initialized value here.

It might be tricky to come up with a test case that triggers the incorrect behavior with the code as-written, but one way you might be able to do it is by fiddling with the order of declarations in your Sema test. Try calling the memberwise init before the struct declaration in the file, and it might be enough to trigger this code path before the lazy variable is completed (which happens late).

Reply:
The reason for this check is because lazy variables whose accessors have been synthesized have removed the `hasStorage` check from the variable as well which is an assertion in `isDefaultInitializable`. Those that have not been completed still pass the following check because they are still considered to have their storage.

Reply:
I think its less error-prone to handle both cases uniformly by only checking if `hasStorage()` on properties that are not lazy. If the property has a LazyAttr, you proceed regardless.

Reply:
We don't really want to just check `hasStorage` either because variables without initial expressions proceed and we want to return early if that's not the case. Something like:
```cpp
if (!var->getAttrs().hasAttribute<LazyAttr>() &&
    !var->getParentPatternBinding()->isDefaultInitializable())
  return;
```
could work maybe?

Reply:
Yes, that looks fine.